### PR TITLE
Add compile time tex math support

### PR DESCRIPTION
**Pull request by KendrickAng**
**What is the purpose of this pull request? (put "X" next to an item, remove the rest)**

• [ ] Documentation update
• [ ] Bug fix
• [X ] New feature
• [ ] Enhancement to an existing feature
• [ ] Other, please explain:

**What is the rationale for this request?**
This pull request addresses issue #1313 , to add additional compile-time tex math support to Markbind.

**What changes did you make? (Give an overview)**
This PR adds two new packages to markbind-core: `katex` and `markdown-it-texmath`. `markdown-it-texmath` requires `katex` as a dependency. A new dev-dependency `file-loader` is also added, since `katex`'s css depends on font files `.woff2`.

Some of the user guide changes are shown below:
1. In the landing page
image

2. Under `Formatting Contents`:
image

**Provide some example code that this change will affect:**

```js
// In core-web/src/styles/index.css
@import "markbind.css";
@import "page-nav.css";
@import "site-nav.css";
@import "~katex/dist/katex.css";
@import "~markdown-it-texmath/css/texmath.css";

// In ComponentParser.js
  static _parseAnnotationAttributes(node) {
    if (!_.has(node.attribs, 'v-pre')) {
      node.attribs['v-pre'] = true;
    }
  }
```

**Is there anything you'd like reviewers to focus on?**
The quality of documentation in the User Guide

**Testing instructions:**
1. Run `npm run build:web` to rebuild `markbind.min.css` and `markbind.min.js`.
2. In any markbind website, insert use `\( ... \)` for inline equations and `\[ ... \]` for block-level equations. E.g `\(1+1=2\)`
3. To test edge cases with `{{` which conflict with nunjucks parsing, try this: `\(e^{{\frac{1}{3}} + 1}\)`

**Proposed commit message: (wrap lines at 72 characters)**
Add compile time tex math support

### CODE REVIEW

File: packages/core/src/lib/markdown-it/index.js [hunk 1]
```diff
@@ -14,6 +14,7 @@ markdownIt.use(require('markdown-it-mark'))
   .use(require('markdown-it-table-of-contents'))
   .use(require('markdown-it-task-lists'), {enabled: true})
   .use(require('markdown-it-linkify-images'), {imgClass: 'img-fluid'})
+  .use(require('markdown-it-texmath'), {engine: require('katex'), delimiters: 'dollars'})
```
Comment (ang-zeyu) [inline packages/core/src/lib/markdown-it/index.js:17]: quick comment:
let's use `brackets` instead, since we're planning to use `$$` for underline
It should also be the more conventional format I think

Comment (KendrickAng) [reply packages/core/src/lib/markdown-it/index.js:17]: I'll do that!

Comment (KendrickAng) [reply packages/core/src/lib/markdown-it/index.js]: @ang-zeyu done!

File: packages/core-web/src/styles/index.css [hunk 2]
```diff
@@ -1,3 +1,5 @@
 @import "markbind.css";
 @import "page-nav.css";
 @import "site-nav.css";
+@import url("~katex/dist/katex.css");
```
Comment (ang-zeyu) [inline packages/core-web/src/styles/index.css:4]: what's the additional `url` for?

File: docs/userGuide/syntax/mathformulae.mbdf [hunk 3]
```diff
@@ -0,0 +1,60 @@
+## Math Formulae
+Markbind supports typesetting math equations with [KaTeX](https://katex.org).
+
+<box type="info">
+
+Insert inline equations by enclosing them in round brackets `\( ... \)`.
+
+Insert block equations by enclosing them in square brackets `\[ ... \]`.
```
Comment (ang-zeyu) [inline docs/userGuide/syntax/mathformulae.mbdf:8]: Should this be `display equations` instead? It seems to be referred in that manner quite often.

File: docs/userGuide/syntax/mathformulae.mbdf [hunk 4]
```diff
@@ -0,0 +1,60 @@
+## Math Formulae
+Markbind supports typesetting math equations with [KaTeX](https://katex.org).
```
Comment (ang-zeyu) [inline docs/userGuide/syntax/mathformulae.mbdf:2]: Since `katex` is the rendering engine, not the spec, let's start with the spec first, then list the engine's use

File: docs/userGuide/syntax/mathformulae.mbdf [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+Insert block equations by enclosing them in square brackets `\[ ... \]`.
+
+Insert numbered block equations by enclosing the equation square brackets and the equation number in curly brackets `\[ ... \] (1)`.
+</box>
+
+<box type="important">
+
+You may leave whitespace between the brackets and your equation.
+
+Correct: `\( 1+1=2 \)`, Correct: `\(1+1=2\)`
```
Comment (ang-zeyu) [inline docs/userGuide/syntax/mathformulae.mbdf:17]: is this a specific trait of `markdown-it-texmath` or part of the tex math? I think we can leave this out and assume the reader already know the spec well if its the latter

File: docs/userGuide/syntax/mathformulae.mbdf [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+
+Ensure that there is a newline between an inline equation and a block equation, or formatting will not work.
```
Comment (ang-zeyu) [inline docs/userGuide/syntax/mathformulae.mbdf:19]: it should be not be preceded by a markdown paragraph.
Should we remove this since its a markdown spec thing instead? Then enhance the examples further below to make the use case of inline / display equations clearer instead

e.g. this does not work as well
```
abc
\[2+2=4\]

or this 

> abc
> \[2+2=4\]
```

File: docs/userGuide/syntax/mathformulae.mbdf [hunk 7]
```diff
@@ -0,0 +1,60 @@
+## Math Formulae
```
Comment (ang-zeyu) [inline docs/userGuide/syntax/mathformulae.mbdf:1]: I've never used latex much before, so take these with a pinch of salt 😐

File: packages/core-web/src/styles/index.css (hunk 2 above)
Comment (ang-zeyu) [reply packages/core-web/src/styles/index.css:4]: Untitled

You'll need to copy the font files to the appropriate output directory too

Comment (KendrickAng) [reply packages/core-web/src/styles/index.css:4]: The additional `url` was how I used to import other css files. Upon checking https://developer.mozilla.org/en-US/docs/Web/CSS/@import, it seems that `url` is redundant. I'll remove it.

File: docs/userGuide/syntax/mathformulae.mbdf (hunk 5 above)
Comment (KendrickAng) [reply docs/userGuide/syntax/mathformulae.mbdf:17]: Whitespaces being allowed seems to be default behaviour for Tex (based on short experiments only on Overleaf). I had added that in previously since using `$` as delimiters didn't allow whitespace, but with the change in delimiters I suppose it would be better to remove it.

File: packages/core-web/src/styles/index.css (hunk 2 above)
Comment (KendrickAng) [reply packages/core-web/src/styles/index.css]: image
I've fixed the issue, added extra logic in Site to copy over the files to the appropriate directory! (imo, that was more straightforward compared to two different webpack configurations for dev and prod)

**Comment by KendrickAng** [comment]
@ang-zeyu I've made the changes to the user guide and Site.js based on your comments!

File: docs/index.md [hunk 8]
```diff
@@ -61,6 +61,30 @@ An example that uses GFMD synatx for task lists:
...
+
+```markdown
+Euler's Identity \(e^{i\pi}+1=0\) is a beautiful formula.
+```
+  </td>
+  <td>&nbsp;→&nbsp;</td>
+  <td><br>
+<tip-box>
+
+Euler's Identity \(e^{i\pi}+1=0\) is a beautiful formula.
```
Comment (crphang) [inline docs/index.md:80]: A little curious whether an expression like `e^{{\frac{1}_{3}} + 1}` will cause nunjucks to misbehave in Markbind now. Do we need this: https://markbind.org/userGuide/tipsAndTricks.html#using-raw-endraw-to-display-content?

Comment (KendrickAng) [reply docs/index.md:80]: image
It does seems to be causing nunjucks rendering issues. I'm not too familiar with the intricacies nunjucks (like how to allow for `{{` here), perhaps @ang-zeyu can advise?

Comment (ang-zeyu) [reply docs/index.md:80]: we allow nunjucks syntax anywhere, even in seemingly unintuitive places like fenced code blocks (we even use this in our ug), the downside of this flexibility is that `{% raw %} {% endraw %}` will be necessary to resolve any ambiguity.

since this seems like a common use case (again correct me if im wrong, havent used latex much before), maybe you can add a warning `<box>` to the ug per @crphang's observation!

`v-pre` shouldn't be though, versus using a client side tex math solution like mathjax, since the equations are spit out at compile time

File: docs/userGuide/syntax/mathformulae.mbdf [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+<box type="info">
+
+Insert `inline` equations by enclosing them in round brackets `\( ... \)`.
+
+Insert `display` equations by enclosing them in square brackets `\[ ... \]`.
+
+Insert numbered `display` equations by enclosing the equation square brackets and the equation number in curly brackets `\[ ... \] (1)`.
+</box>
+
+<box type="important">
```
Comment (ang-zeyu) [inline docs/userGuide/syntax/mathformulae.mbdf:13]: should we move this to the bottom? since the warning would make more sense after presenting the feature

File: docs/userGuide/syntax/mathformulae.mbdf [hunk 10]
```diff
@@ -0,0 +1,99 @@
+## Math Formulae
+Markbind supports typesetting TeX math equations. [KaTeX](https://katex.org) is used as a fast math renderer.
+
+<box type="info">
+
+Insert `inline` equations by enclosing them in round brackets `\( ... \)`.
```
Comment (ang-zeyu) [inline docs/userGuide/syntax/mathformulae.mbdf:6]: would it better to bold the **inline/display** or link it somewhere? let's use the inline code syntax just for code

File: docs/userGuide/syntax/mathformulae.mbdf [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+
+Insert numbered `display` equations by enclosing the equation square brackets and the equation number in curly brackets `\[ ... \] (1)`.
+</box>
+
+<box type="important">
+
+When typing `display` equations, this does not work as well:
+
+```markdown
+abc
```
Comment (ang-zeyu) [inline docs/userGuide/syntax/mathformulae.mbdf:18]: https://github.com/MarkBind/markbind/pull/1335#discussion_r480207052

sorry if it was misinterpreted, I didn't mean for you to place this exactly in the docs, it was only to illustrate that it is intentional behaviour by the [commonmark spec](https://commonmark.org/)

I think we should remove _all_ problems related to the markdown spec(s) in general, and let the author get accustomed to it.
if not, just brief mentions that universally apply, like the one we have for blank lines here https://markbind.org/userGuide/usingHtmlJavaScriptCss.html, otherwise every section in our ug would be double in size =X

File: docs/userGuide/syntax/mathformulae.mbdf [hunk 12]
```diff
@@ -0,0 +1,99 @@
...
+```markdown
+> abc
+> \[2+2=4\]
+```
+
+If your equation requires special Nunjucks tags like {% raw %}`{{`{% endraw %} or {% raw %}`}}`{% endraw %},use a
+[raw-endraw block together with v-pre](https://markbind.org/userGuide/tipsAndTricks.html#using-raw-endraw-to-display-content):
+
+```markdown
+{% raw %}\(e^{{\frac{1}{3}} + 1}\){% endraw %} {v-pre}
```
Comment (ang-zeyu) [inline docs/userGuide/syntax/mathformulae.mbdf:32]: as mentioned `v-pre` _shouldn't_ be necessary, which is one of the benefits of doing it at compile time to. do double check though

File: docs/userGuide/syntax/mathformulae.mbdf [hunk 13]
```diff
@@ -0,0 +1,99 @@
...
+
+\( e^{i\pi}+1=0 \)
+\(\frac{1}{3}+\frac{2}{3}=1\)
+
+\[ e^{i\pi}+1=0 \]
+\[\frac{1}{3}+\frac{2}{3}=1\] (1)
+
+{% raw %}\(e^{{\frac{1}{3}} + 1}\){% endraw %} {v-pre}
+
+Bad examples:
```
Comment (ang-zeyu) [inline docs/userGuide/syntax/mathformulae.mbdf:92]: let's remove this too,
instead make the examples directly above here resemble actual use cases as best; i.e. some _short_ text around an inline equation, then present another relavant equation that is better highlighted in display format

File: docs/userGuide/syntax/mathformulae.mbdf [hunk 14]
```diff
@@ -0,0 +1,99 @@
+## Math Formulae
+Markbind supports typesetting TeX math equations. [KaTeX](https://katex.org) is used as a fast math renderer.
+
+<box type="info">
```
Comment (ang-zeyu) [inline docs/userGuide/syntax/mathformulae.mbdf:4]: let's present this vital piece of information about the feature in plain markup, without the box which readers might just skip over

**Review by ang-zeyu — COMMENTED** [review]
implementation looks good, thanks! : )

just a few more docs nits:

File: docs/index.md (hunk 8 above)
Comment (crphang) [reply docs/index.md:80]: When working on nunjuck-escaped special tags with @ang-zeyu, latex's syntax was one of the consideration. Maybe this can be wrapped in nunjuck-escaped special tags? You can check out puml plugin if it helps!

Comment (ang-zeyu) [reply docs/index.md:80]: > nunjuck-escaped special tags

@crphang this hasen't been implemented yet, I don't think we should either, theres several cases where its useful to have nunjucks syntax complement the "special tag", e.g. https://github.com/MarkBind/markbind/issues/1256

Comment (ang-zeyu) [reply docs/index.md:80]: wrapping texmath in html may also be rather troublesome (usage wise) to be worth the cost; that said if there is a way to use only the same `\[ \(` syntax that'd be great though (likely needs to be done via patching nunjucks itself)

File: docs/userGuide/syntax/mathformulae.mbdf (hunk 9 above)
Comment (KendrickAng) [reply docs/userGuide/syntax/mathformulae.mbdf:13]: That makes sense; a user would want to know how to do things first before looking at its edge cases

File: docs/userGuide/syntax/mathformulae.mbdf (hunk 11 above)
Comment (KendrickAng) [reply docs/userGuide/syntax/mathformulae.mbdf:18]: ah I see, I'll remove this section instead and assume the reader has read the `usingHtmlJavaScriptCss.html` section

File: docs/userGuide/syntax/mathformulae.mbdf [hunk 15]
```diff
@@ -0,0 +1,99 @@
...
+abc
+\[2+2=4\]
+```
+
+or this
+```markdown
+> abc
+> \[2+2=4\]
+```
+
```
Comment (KendrickAng) [inline docs/userGuide/syntax/mathformulae.mbdf:27]: test comment

File: docs/userGuide/syntax/mathformulae.mbdf (hunk 12 above)
Comment (KendrickAng) [reply docs/userGuide/syntax/mathformulae.mbdf:32]: I'm not exactly sure why, but leaving out the {v-pre} causes a blank page to show:
image

Whereas leaving it in allows everything to load correctly, so I left it in for now.

**Comment by KendrickAng** [comment]
@ang-zeyu, I've updated the user guide based on your comments. How does it look now?

File: docs/userGuide/syntax/mathformulae.mbdf (hunk 12 above)
Comment (ang-zeyu) [reply docs/userGuide/syntax/mathformulae.mbdf]: seems like katex also spits out MathML accessibility output retaining the original equation by default:

from view page source:
```
<semantics>
--
  | <mrow>
  | <msup>
  | <mi>e</mi>
  | <mrow>
  | <mfrac>
  | <mn>1</mn>
  | <mn>3</mn>
  | </mfrac>
  | <mo>+</mo>
  | <mn>1</mn>
  | </mrow>
  | </msup>
  | </mrow>
  | <annotation encoding="application/x-tex">e^{{\frac{1}{3}} + 1}</annotation>
  | </semantics>

```

let's auto add the `v-pre` to this in `componentParser`

```
<annotation encoding="application/x-tex">e^{{\frac{1}{3}} + 1}</annotation>
```

**Comment by ang-zeyu** [comment]
> @ang-zeyu, I've updated the user guide based on your comments. How does it look now?

really neat now! let's auto add in the `v-pre` to `<annotation encoding="application/x-tex">` which should be only a few lines in `ComponentParser`, then remove it from the ug, and we can merge this in

**Comment by KendrickAng** [comment]
Done!

**Review by ang-zeyu — APPROVED** [review]
